### PR TITLE
Support expressions in SQL generator functions

### LINKED ISSUE #2930: Type coercion for UDFs should not implicitly convert VARCHAR to INT
This query works, but should throw something like `Cannot apply VARCHAR to INT argument`:
```sql
SELECT * FROM TABLE(generate_series('0', '5'))
```

### PR DESCRIPTION
**Pull request by gierlachg**
Added support for expressions (but not for scalar subqueries as those are not supported at all in current engine) to SQL generators (`generate_series` & `generate_stream`).

Adding support for expressions to file table functions will be addressed in a separate PR.

Fixes #2930

### CODE REVIEW

**Review by viliam-durina — APPROVED** [review]

**Comment by viliam-durina** [comment]
The build will pass after hazelcast/hazelcast#18328 is merged

File: hazelcast-jet-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/logical/FullFunctionScanLogicalRules.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+            RexVisitor<Expression<?>> visitor = OptUtils.createRexToExpressionVisitor(FailingFieldTypeProvider.INSTANCE);
+            List<Object> arguments = IntStream.range(0, specificFunction.getOperandCountRange().getMax())
+                    .mapToObj(index -> {
+                        if (index < operands.size()) {
+                            Expression<?> expression = operands.get(index).accept(visitor);
+                            return expression.eval(EmptyRow.INSTANCE, ExpressionUtil.NOT_IMPLEMENTED_ARGUMENTS_CONTEXT);
+                        } else {
+                            return null;
+                        }
+                    }).collect(toList());
```
[the remark this answers is not in the document]
Comment (gierlachg) [reply hazelcast-jet-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/logical/FullFunctionScanLogicalRules.java:69]: Not sure, then each variadic function would have to deal with different number of arguments on its own (apart from handling `null`s which it has to do anyway), this way it's handled in single place.

[the remark this answers is not in the document]
Comment (viliam-durina) [reply hazelcast-jet-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/logical/FullFunctionScanLogicalRules.java:69]: As you wish, both approaches are equivalent, it's about style.

**Comment by gierlachg** [comment]
run-lab-run

**Pull request merged**